### ISSUE TITLE
[LightGBM] [Fatal] [tweedie]: at least one target label is negative

### ISSUE THREAD

**Issue by barrettlayman** [tidymodels/bonsai#50]
I receive an error while using Bonsai to train a LightGBM model with a tweedie objective.

The same error does not occur when I train a model using LightGBM without Bonsai.

I have included an example below where I replace one record in the Boston dataset with a zero.

Thank you!

```r
library(lightgbm)
#> Loading required package: R6
library(caret)
#> Loading required package: ggplot2
#> Loading required package: lattice
library(bonsai)
#> Warning: package 'bonsai' was built under R version 4.2.1
#> Loading required package: parsnip

boston <- MASS::Boston
str(boston)
#> 'data.frame':    506 obs. of  14 variables:
#>  $ crim   : num  0.00632 0.02731 0.02729 0.03237 0.06905 ...
#>  $ zn     : num  18 0 0 0 0 0 12.5 12.5 12.5 12.5 ...
#>  $ indus  : num  2.31 7.07 7.07 2.18 2.18 2.18 7.87 7.87 7.87 7.87 ...
#>  $ chas   : int  0 0 0 0 0 0 0 0 0 0 ...
#>  $ nox    : num  0.538 0.469 0.469 0.458 0.458 0.458 0.524 0.524 0.524 0.524 ...
#>  $ rm     : num  6.58 6.42 7.18 7 7.15 ...
#>  $ age    : num  65.2 78.9 61.1 45.8 54.2 58.7 66.6 96.1 100 85.9 ...
#>  $ dis    : num  4.09 4.97 4.97 6.06 6.06 ...
#>  $ rad    : int  1 2 2 3 3 3 5 5 5 5 ...
#>  $ tax    : num  296 242 242 222 222 222 311 311 311 311 ...
#>  $ ptratio: num  15.3 17.8 17.8 18.7 18.7 18.7 15.2 15.2 15.2 15.2 ...
#>  $ black  : num  397 397 393 395 397 ...
#>  $ lstat  : num  4.98 9.14 4.03 2.94 5.33 ...
#>  $ medv   : num  24 21.6 34.7 33.4 36.2 28.7 22.9 27.1 16.5 18.9 ...
dim(boston)
#> [1] 506  14

set.seed(12)

indexes <- createDataPartition(boston$medv, p = .85, list = F)
train <- boston[indexes, ]
test <- boston[-indexes, ]

train_x <- train[, -14]
train_x <- scale(train_x)[, ]
train_y <- train[, 14]

test_x <- test[, -14]
test_x <- scale(test[, -14])[, ]
test_y <- test[, 14]

dtrain <- lgb.Dataset(train_x, label = train_y)
dtest <- lgb.Dataset.create.valid(dtrain, test_x, label = test_y)

# define parameters
params <- list(
  objective = "tweedie",
  tweedie_variance_power = 1.5,
  metric = "l2",
  min_data = 1L,
  learning_rate = .3
)

# validataion data
valids <- list(test = dtest)

# train model
model <- lgb.train(
  params = params,
  data = dtrain,
  nrounds = 5L,
  valids = valids
)
#> [LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001566 seconds.
#> You can set `force_row_wise=true` to remove the overhead.
#> And if memory is not enough, you can set `force_col_wise=true`.
#> [LightGBM] [Info] Total Bins 1078
#> [LightGBM] [Info] Number of data points in the train set: 432, number of used features: 13
#> [LightGBM] [Info] Start training from score 3.117960
#> [1] "[1]:  test's l2:35.351"
#> [1] "[2]:  test's l2:21.9136"
#> [1] "[3]:  test's l2:15.9518"
#> [1] "[4]:  test's l2:13.4712"
#> [1] "[5]:  test's l2:12.782"

# Test tweedie with a zero in target variable

boston <- MASS::Boston

boston[, 14][[1]] <- 0

str(boston)
#> 'data.frame':    506 obs. of  14 variables:
#>  $ crim   : num  0.00632 0.02731 0.02729 0.03237 0.06905 ...
#>  $ zn     : num  18 0 0 0 0 0 12.5 12.5 12.5 12.5 ...
#>  $ indus  : num  2.31 7.07 7.07 2.18 2.18 2.18 7.87 7.87 7.87 7.87 ...
#>  $ chas   : int  0 0 0 0 0 0 0 0 0 0 ...
#>  $ nox    : num  0.538 0.469 0.469 0.458 0.458 0.458 0.524 0.524 0.524 0.524 ...
#>  $ rm     : num  6.58 6.42 7.18 7 7.15 ...
#>  $ age    : num  65.2 78.9 61.1 45.8 54.2 58.7 66.6 96.1 100 85.9 ...
#>  $ dis    : num  4.09 4.97 4.97 6.06 6.06 ...
#>  $ rad    : int  1 2 2 3 3 3 5 5 5 5 ...
#>  $ tax    : num  296 242 242 222 222 222 311 311 311 311 ...
#>  $ ptratio: num  15.3 17.8 17.8 18.7 18.7 18.7 15.2 15.2 15.2 15.2 ...
#>  $ black  : num  397 397 393 395 397 ...
#>  $ lstat  : num  4.98 9.14 4.03 2.94 5.33 ...
#>  $ medv   : num  0 21.6 34.7 33.4 36.2 28.7 22.9 27.1 16.5 18.9 ...
dim(boston)
#> [1] 506  14

set.seed(12)

indexes <- createDataPartition(boston$medv, p = .85, list = F)
train <- boston[indexes, ]
test <- boston[-indexes, ]

train_x <- train[, -14]
train_x <- scale(train_x)[, ]
train_y <- train[, 14]

test_x <- test[, -14]
test_x <- scale(test[, -14])[, ]
test_y <- test[, 14]

dtrain <- lgb.Dataset(train_x, label = train_y)
dtest <- lgb.Dataset.create.valid(dtrain, test_x, label = test_y)

# define parameters
params <- list(
  objective = "tweedie",
  tweedie_variance_power = 1.5,
  metric = "l2",
  min_data = 1L,
  learning_rate = .3
)

# validataion data
valids <- list(test = dtest)

# train model
model <- lgb.train(
  params = params,
  data = dtrain,
  nrounds = 5L,
  valids = valids
)
#> [LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001375 seconds.
#> You can set `force_row_wise=true` to remove the overhead.
#> And if memory is not enough, you can set `force_col_wise=true`.
#> [LightGBM] [Info] Total Bins 1071
#> [LightGBM] [Info] Number of data points in the train set: 432, number of used features: 13
#> [LightGBM] [Info] Start training from score 3.115550
#> [1] "[1]:  test's l2:72.8958"
#> [1] "[2]:  test's l2:94.8558"
#> [1] "[3]:  test's l2:92.4295"
#> [1] "[4]:  test's l2:91.6841"
#> [1] "[5]:  test's l2:95.9142"

bonsai_tweedie_test <-
  boost_tree() %>%
  set_engine(
    engine = "lightgbm",
    objective = "tweedie",
    tweedie_variance_power = 1.7,
    metric = "l2"
  ) %>%
  set_mode(mode = "regression") %>%
  fit(
    formula = medv ~ .,
    data = boston
  )
#> Error in try({ : [tweedie]: at least one target label is negative
#> Error in initialize(...): lgb.Booster: cannot create Booster handle
```

**Comment by simonpcouch** [comment]
Thanks for the issue! This is now good to go in the dev version:

``` r
library(bonsai)
#> Loading required package: parsnip
library(MASS)

data(Boston)

boston <- MASS::Boston
boston[, 14][1] <- 0

boost_tree() %>%
  set_engine(
    engine = "lightgbm",
    objective = "tweedie",
    tweedie_variance_power = 1.7,
    metric = "l2"
  ) %>%
  set_mode(mode = "regression") %>%
  fit(
    formula = medv ~ .,
    data = boston
  )
#> parsnip model object
#> 
#> <lgb.Booster>
#>   Public:
#>     add_valid: function (data, name) 
#>     best_iter: -1
#>     best_score: NA
#>     current_iter: function () 
#>     dump_model: function (num_iteration = NULL, feature_importance_type = 0L) 
#>     eval: function (data, name, feval = NULL) 
#>     eval_train: function (feval = NULL) 
#>     eval_valid: function (feval = NULL) 
#>     finalize: function () 
#>     initialize: function (params = list(), train_set = NULL, modelfile = NULL, 
#>     lower_bound: function () 
#>     params: list
#>     predict: function (data, start_iteration = NULL, num_iteration = NULL, 
#>     raw: NA
#>     record_evals: list
#>     reset_parameter: function (params, ...) 
#>     rollback_one_iter: function () 
#>     save: function () 
#>     save_model: function (filename, num_iteration = NULL, feature_importance_type = 0L) 
#>     save_model_to_string: function (num_iteration = NULL, feature_importance_type = 0L) 
#>     set_train_data_name: function (name) 
#>     to_predictor: function () 
#>     update: function (train_set = NULL, fobj = NULL) 
#>     upper_bound: function () 
#>   Private:
#>     eval_names: NULL
#>     get_eval_info: function () 
#>     handle: lgb.Booster.handle
#>     higher_better_inner_eval: NULL
#>     init_predictor: NULL
#>     inner_eval: function (data_name, data_idx, feval = NULL) 
#>     inner_predict: function (idx) 
#>     is_predicted_cur_iter: list
#>     name_train_set: training
#>     name_valid_sets: list
#>     num_class: 1
#>     num_dataset: 1
#>     predict_buffer: list
#>     set_objective_to_none: FALSE
#>     train_set: lgb.Dataset, R6
#>     train_set_version: 1
#>     valid_sets: list
```

<sup>Created on 2022-08-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by barrettlayman** [comment]
Thank you!  I just checked my code after I installed the Dev version.  Everything ran as expected.  Appreciate the quick response.
